solver/llbsolver/provenance/predicate.go

```diff
@@ -83,6 +83,22 @@ func slsaMaterials(srcs Sources) ([]slsa.ProvenanceMaterial, error) {
 		out = append(out, material)
 	}
 
+	for _, s := range srcs.ImageBlobs {
+		uri, err := purl.RefToPURL("docker-blob", s.Ref, nil)
```

Comment:
Surfacing from Slack:

> Am I correct in that you’re inventing your own docker-blob PURL scheme?
> I wouldn't invent a new PURL ecosystem; they're well-defined in https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst; pkg:docker/ with a specifier (e.g. ?buildkit_type=blob) makes more sense to me.

@tonistiigi 
> I don’t know what else to put there. I’m extending the docker scheme because it is the same protocol. But I already have a problem that I want to be able to point to a local OCI layout this way as well that would need another name.

---

I'd like to suggest some alternatives as I don't think inventing our own scheme outside the PURL spec is a good idea:

```
# Embedding the original source as a URL + checksum:
pkg:generic/openssl@1.1.10g?download_url=https://openssl.org/source/openssl-1.1.0g.tar.gz&checksum=sha256:de4d501267da&buildkit_blob=docker.io/mynamespace/myrepo@sha256:...
# Extending the existing docker scheme:
pkg:docker/mynamespace/myrepo@sha256%3A244fd47e07d10?buildkit_type=blob
```

Reply:
Tonis and I discussed this further; what likely makes more sense is passing the provenance all the way through (so, use the existing URL and hash even when doing a "replay" from "registry mirrored inputs"; and have some BuildKit-specific sourcepolicy-map file that records what verifiable mirrored input was used for the specific build).

Reply:
I think it's mostly because "try a digest as a manifest and fall back to trying it as a blob" isn't an existing behavior and doesn't really make sense in the general case; there are separate APIs to get this content (for better and for worse) and having a hint makes it more clear what is actually expected/prevents fragmentation among tools that return a 404 because they didn't fall back to blob.

This is all mostly hypothetical as this is intended to be an internal implementation detail of "replayable" builds, but nonetheless PURLs are meant to be interoperable and commonly understood...

If we use the latter type, we should probably PR it back to the PURL spec.

Reply:
https://github.com/package-url/purl-spec/blob/5b15089c476d35b7facccaee47357df4f2b47fcb/PURL-TYPES.rst?plain=1#L211-L224
```rst
docker
------
``docker`` for Docker images:

- The default repository is ``https://hub.docker.com``.
- The ``namespace`` is the registry/user/organization if present.
- The version should be the image id sha256 or a tag. Since tags can be moved,
  a sha256 image id is preferred.
- Examples::

      pkg:docker/cassandra@latest
      pkg:docker/smartentry/debian@dc437cc87d10
      pkg:docker/customer/dockerimage@sha256%3A244fd47e07d10?repository_url=gcr.io

```

https://github.com/package-url/purl-spec/blob/5b15089c476d35b7facccaee47357df4f2b47fcb/PURL-TYPES.rst?plain=1#L416-L444

```rst
oci
------------
``oci`` for all artifacts stored in registries that conform to the
`OCI Distribution Specification <https://github.com/opencontainers/distribution-spec>`_,
including container images built by Docker and others:

- There is no canonical package repository for OCI artifacts. Therefore
  ``oci`` purls must be registry agnostic by default. To specify the repository,
  provide a ``repository_url`` value.
- OCI purls do not contain a ``namespace``, although, ``repository_url`` may
  contain a namespace as part of the physical location of the package.
- The ``name`` is not case sensitive and must be lowercased. The name is the
  last fragment of the repository name. For example if the repository
  name is ``library/debian`` then the ``name`` is ``debian``.
- The ``version`` is the ``sha256:hex_encoded_lowercase_digest`` of the
  artifact and is required to uniquely identify the artifact.
- Optional qualifiers may include:

  - ``arch``: key for a package architecture, when relevant.
  - ``repository_url``: A repository URL where the artifact may be found, but not
    intended as the only location. This value is encouraged to identify a
    location the content may be fetched.
  - ``tag``: artifact tag that may have been associated with the digest at the time.
- Examples::

      pkg:oci/debian@sha256%3A244fd47e07d10?repository_url=docker.io/library/debian&arch=amd64&tag=latest
      pkg:oci/debian@sha256%3A244fd47e07d10?repository_url=ghcr.io/debian&tag=bullseye
      pkg:oci/static@sha256%3A244fd47e07d10?repository_url=gcr.io/distroless/static&tag=latest
      pkg:oci/hello-wasm@sha256%3A244fd47e07d10?tag=v1
```


Since a blob is not an "image", using `oci` type may make more sense.

`?ref_type=blob` doesn't seem needed.